datafusion/expr/src/expr.rs

```diff
@@ -234,12 +236,33 @@ pub struct BinaryExpr {
     pub op: Operator,
     /// Right-hand side of the expression
     pub right: Box<Expr>,
+    /// The data type of the expression, if known
```

Comment:
I don't understand the reason for this change -- is it no longer possible to calculate the output type of a `BinaryExpr` from its inputs? 

If the desired output type is different then couldn't we `cast` the expr to the type?